src/basic/fileio.c

```diff
@@ -1278,3 +1254,97 @@ int fputs_with_space(FILE *f, const char *s, const char *separator, bool *space)
 
         return fputs(s, f);
 }
+
+int open_tmpfile_unlinkable(const char *directory, int flags) {
+        char *p;
+        int fd;
+
+        assert(directory);
+
+        /* Returns an unlinked temporary file that cannot be linked into the file system anymore */
+
+#ifdef O_TMPFILE
+        /* Try O_TMPFILE first, if it is supported */
+        fd = open(directory, flags|O_TMPFILE|O_EXCL, S_IRUSR|S_IWUSR);
+        if (fd >= 0)
+                return fd;
+#endif
+
+        /* Fall back to unguessable name + unlinking */
+        p = strjoina(directory, "/systemd-tmp-XXXXXX");
+
+        fd = mkostemp_safe(p, flags);
+        if (fd < 0)
+                return fd;
+
+        (void) unlink(p);
+
+        return fd;
+}
+
+int open_tmpfile_linkable(const char *target, int flags, char **ret_path) {
+        _cleanup_free_ char *tmp = NULL;
+        int r, fd;
+
+        assert(target);
+        assert(ret_path);
+
+        /* Creates a temporary file, that shall be renamed to "target" later. If possible, this uses O_TMPFILE – in
+         * which case "ret_path" will be returned as NULL. If not possible a the tempoary path name used is returned in
+         * "ret_path". Use link_tmpfile() below to rename the result after writing the file in full. */
+
+#ifdef O_TMPFILE
+        {
+                _cleanup_free_ char *dn = NULL;
+
+                dn = dirname_malloc(target);
+                if (!dn)
+                        return -ENOMEM;
+
+                fd = open(dn, O_TMPFILE|flags, 0640);
+                if (fd >= 0) {
+                        *ret_path = NULL;
+                        return fd;
+                }
+
+                log_debug_errno(errno, "Failed to use O_TMPFILE on %s: %m", dn);
+        }
+#endif
+
+        r = tempfn_random(target, NULL, &tmp);
+        if (r < 0)
+                return r;
+
+        fd = open(tmp, O_CREAT|O_EXCL|O_NOFOLLOW|O_NOCTTY|flags, 0640);
+        if (fd < 0)
+                return -errno;
+
+        *ret_path = tmp;
+        tmp = NULL;
+
+        return fd;
+}
+
+int link_tmpfile(int fd, const char *path, const char *target) {
```

Comment:
`rename`
> If newpath already exists, it will be atomically replaced

`linkat`
> If newpath exists, it will not be overwritten.

So we get different results on the different filesystems:

`link_tmpfile` successfully overwrites an existing file on `reiserfs`:
```
$ findmnt /reiserfs
TARGET    SOURCE       FSTYPE   OPTIONS
/reiserfs /dev/loop0p1 reiserfs rw,relatime

$ touch /reiserfs/tmpfile

$ strace ./test-linkat /reiserfs/tmpfile
access("/reiserfs/tmpfile", F_OK)       = 0
open("/reiserfs", O_RDWR|O_DIRECTORY|O_CLOEXEC|O_TMPFILE) = -1 EOPNOTSUPP (Operation not supported)
writev(2, [{"Failed to use O_TMPFILE on /reis"..., 61}, {"\n", 1}], 2Failed to use O_TMPFILE on /reiserfs: Operation not supported
) = 62
getrandom("G\326\357\342\312\303p'", 8, GRND_NONBLOCK) = 8
open("/reiserfs/.#tmpfile746dfe2eac3c0772", O_RDWR|O_CREAT|O_EXCL|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC, 0640) = 3
write(3, "foobar\n", 7)                 = 7
rename("/reiserfs/.#tmpfile746dfe2eac3c0772", "/reiserfs/tmpfile") = 0
open("/reiserfs/tmpfile", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0640, st_size=7, ...}) = 0
read(4, "foobar\n", 4096)               = 7
close(4)                                = 0
```

`link_tmpfile` fails on `xfs`:
```
$ findmnt /xfs
TARGET SOURCE       FSTYPE OPTIONS
/xfs   /dev/loop0p1 xfs    rw,relatime,attr2,inode64,noquota

$ touch /xfs/tmpfile

$ strace ./test-linkat /xfs/tmpfile
...
access("/xfs/tmpfile", F_OK)            = 0
open("/xfs", O_RDWR|O_DIRECTORY|O_CLOEXEC|O_TMPFILE) = 3
write(3, "foobar\n", 7)                 = 7
linkat(AT_FDCWD, "/proc/self/fd/3", AT_FDCWD, "/xfs/tmpfile", AT_SYMLINK_FOLLOW) = -1 EEXIST (File exists)
```
It doesn't really matter in the `systemd-coredump`. But I'm not sure about common library code.